include/Messenger.h

```diff
@@ -0,0 +1,93 @@
+/*
+ * Messenger: class to abstract the routing of Open Sound Control messages from the core
+ *   to another area of the core and/or the gui
+ *
+ * (c) 2015 Colin Wallace (https://github.com/Wallacoloo)
+ *
+ * This file is part of LMMS - http://lmms.io
+ *
+ * This file is released under the MIT License.
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ * 
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ * 
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#ifndef MESSENGER_H
+#define MESSENGER_H
+
+#include <cstring>
+
+#include <QReadWriteLock>
+#include <QSet>
+
+class QString;
+
+class OscMsgListener;
+
+
+class Messenger
+{
+public:
+	// const strings to specify the Open Sound Control endpoint locations,
+	// e.g. "/mixer/ch/n/volume"
+	class Endpoints
+	{
+	public:
+		// Endpoints to send general and already-translated warnings / error messages
+		static const char *Warning;
+		static const char *Error;
+
+		static const char *WaveTableInit;
+		static const char *MixerDevInit;
+		static const char *MixerProcessingStart;
+	};
+
+	// Send message to indicate that the following components have completed their initialization
+	void broadcastWaveTableInit();
```

Comment:
@michaelgregorius I intended to keep those methods going forward. The most important thing is to keep string constants related to OSC (i.e. the endpoint names) from being defined in more than one file, as mismatched strings cannot be detected as a compile-time error. Hence the `Endpoints` class.

Secondly, any library-specific OSC interactions should be limited to as few files as possible - I don't want the engine to have to encode its own OSC messages, for example. That essentially leaves two options:

1. Declare the Warning and Error broadcast functions and then a single `InitComponent(QString endpoint)` function.
2. Declare the Warning and Error broadcast functions and then another `InitComponentX()` function for each possible component.

I chose no.2 so that each endpoint would correspond to exactly one broadcast function. One could make an argument for no.1 on the basis that it further decouples this file from the lmms internals. If so, that also requires that the string constants be declared in each component's header rather than Messenger.h, otherwise you're still making assumptions about lmms internals. My argument for no.2 is that this gives you all the documentation for the communications interface in one single location, so you don't have to go hunting down for a list of endpoints that's spread out through the whole code base when it comes to interfacing with lmms.